source/agora/test/Base.d

```diff
@@ -1315,6 +1315,10 @@ public struct TestConf
 
     /// How often blocks should be created - in seconds
     uint block_interval_sec = 1;
+
+    /// Define some Byzantine nodes
+    size_t byzantine_not_signing_count = 0;
```

Comment:
This looks like leaky abstractions to me. We added a config option in `Base` but it's only used in a derived module.

Why not make these two options templated on the `ByzantineManager`? Something like:

```diff
diff --git a/source/agora/test/Byzantine.d b/source/agora/test/Byzantine.d
index 3b6b72ed0..0ff89c7a9 100644
--- a/source/agora/test/Byzantine.d
+++ b/source/agora/test/Byzantine.d
@@ -43,7 +43,7 @@ import std.stdio;
 import core.exception;
 import core.stdc.time;
 
-enum ByzantineReason 
+enum ByzantineReason
 {
     none,
     not_signing_envelope,
@@ -53,7 +53,7 @@ enum ByzantineReason
 extern(C++) class ByzantineNominator : TestNominator
 {
     private ByzantineReason reason;
-    
+
     extern(D) this (immutable(ConsensusParams) params,
         Clock clock, NetworkManager network, KeyPair key_pair, Ledger ledger,
         TaskManager taskman, ByzantineReason reason)
@@ -66,7 +66,7 @@ extern(C++) class ByzantineNominator : TestNominator
     // refuse to sign
     extern(C++) override void signEnvelope (ref SCPEnvelope envelope)
     {
-        switch (reason) 
+        switch (reason)
         {
             bad_signing_envelope:
             {
@@ -122,7 +122,9 @@ class ByzantineNodeBadSigning : TestValidatorNode
 }
 
 /// creates some nodes depending on the testConf which will not sign or will sign with invalid signature
-class ByzantineManager : TestAPIManager
+class ByzantineManager (
+    size_t byzantine_not_signing_count = 0, size_t byzantine_bad_signing_count = 0)
+    : TestAPIManager
 {
     ///
     public this (immutable(Block)[] blocks, TestConf test_conf,
@@ -134,16 +136,16 @@ class ByzantineManager : TestAPIManager
     public override void createNewNode (Config conf,
         string file = __FILE__, int line = __LINE__)
     {
-        if (this.nodes.length < test_conf.byzantine_not_signing_count + test_conf.byzantine_bad_signing_count)
+        if (this.nodes.length < byzantine_not_signing_count + byzantine_bad_signing_count)
         {
             auto time = new shared(time_t)(this.initial_time);
             assert(conf.node.is_validator);
             RemoteAPI!TestAPI node;
-            if (this.nodes.length < test_conf.byzantine_not_signing_count)
+            if (this.nodes.length < byzantine_not_signing_count)
                 node = RemoteAPI!TestAPI.spawn!ByzantineNodeNotSigning(
                 conf, &this.reg, this.blocks, this.test_conf.txs_to_nominate,
                 time, conf.node.timeout);
-            else 
+            else
                 node = RemoteAPI!TestAPI.spawn!ByzantineNodeBadSigning(
                 conf, &this.reg, this.blocks, this.test_conf.txs_to_nominate,
                 time, conf.node.timeout);
@@ -158,9 +160,8 @@ class ByzantineManager : TestAPIManager
 /// Block should be added if we only have 1 out of 3 validators not signing
 unittest
 {
-    TestConf conf = { validators : 4, quorum_threshold : 66,  
-        byzantine_not_signing_count : 1 };
-    auto network = makeTestNetwork!ByzantineManager(conf);
+    TestConf conf = { validators : 4, quorum_threshold : 66 };
+    auto network = makeTestNetwork!(ByzantineManager!(1, 0))(conf);
     network.start();
     scope(exit) network.shutdown();
     scope(failure) network.printLogs();
@@ -175,9 +176,8 @@ unittest
 /// Block should be added if we only have 1 out of 3 validators signing with invalid signature
 unittest
 {
-    TestConf conf = { validators : 4, quorum_threshold : 66 ,  
-        byzantine_bad_signing_count : 1 };
-    auto network = makeTestNetwork!ByzantineManager(conf);
+    TestConf conf = { validators : 4, quorum_threshold : 66 };
+    auto network = makeTestNetwork!(ByzantineManager!(0, 1))(conf);
     network.start();
     scope(exit) network.shutdown();
     scope(failure) network.printLogs();
@@ -189,12 +189,12 @@ unittest
     network.expectBlock(Height(1), 5.seconds);
 }
 
-/// Too many byzantine nodes will prevent block from being externalized 
+/// Too many byzantine nodes will prevent block from being externalized
 unittest
 {
-    TestConf conf = { validators : 4, quorum_threshold : 66, 
+    TestConf conf = { validators : 4, quorum_threshold : 66,
         byzantine_not_signing_count : 1, byzantine_bad_signing_count : 1 };
-    auto network = makeTestNetwork!ByzantineManager(conf);
+    auto network = makeTestNetwork!(ByzantineManager!())(conf);
     network.start();
     scope(exit) network.shutdown();
     scope(failure) network.printLogs();
```

Also please fix your editor so it always strips out trailing spaces on save.

Reply:
Updated file and settings